<img src="https://substack-post-media.s3.amazonaws.com/public/images/730d9518-707f-4263-8925-9679ace2d78f_2102x2327.png" width="1000" height="500"/>


# **Little test**
* https://statinferentielle.fr/test-little-mcar/#:~:text=Hypoth%C3%A8ses%20du%20test%20de%20Little,elles%20sont%20MAR%20ou%20MNAR).

> ## `Hypothèses du test Little`
> Hypothèse nulle (H₀)  : Les données manquantes sont MCAR.
>
> Hypothèse alternative (H₁): Les données ne sont pas MCAR (elles sont MAR ou MNAR).

In [6]:
data = pd.read_csv("Fraud Detection Dataset.csv")

In [7]:
#!pip install pyampute
import pandas as pd
from pyampute.exploration.mcar_statistical_tests import MCARTest
?MCARTest


Init signature: MCARTest(method: str = 'little')
Docstring:     
Statistical hypothesis test for Missing Completely At Random (MCAR)

Performs Little's MCAR test (see `Little, R.J.A. (1988)`_). Null hypothesis: data is Missing Completely At Random (MCAR). Alternative hypothesis: data is not MCAR.

.. _`Little, R.J.A. (1988)`: https://www.tandfonline.com/doi/abs/10.1080/01621459.1988.10478722

Parameters
----------
method : str, {"little", "ttest"}, default : "little"
    Whether to perform a chi-square test on the entire dataset ("little") or separate t-tests for every combination of variables ("ttest"). 

See also
--------
:class:`~pyampute.exploration.md_patterns.mdPatterns` : Displays missing data patterns in incomplete datasets

:class:`~pyampute.ampute.MultivariateAmputation` : Transformer for generating multivariate missingness in complete datasets

Notes
-----
We advise to use Little's MCAR test carefully. Rejecting the null hypothesis may not always mean that data is not MCAR, 

In [8]:
# Source - https://stackoverflow.com/a

import numpy as np
import pandas as pd
from scipy.stats import chi2

def little_mcar_test(data, alpha=0.05):
    """
    Performs Little's MCAR (Missing Completely At Random) test on a dataset with missing values.
    """
    data = pd.DataFrame(data)
    data.columns = ['x' + str(i) for i in range(data.shape[1])]
    data['missing'] = np.sum(data.isnull(), axis=1)
    n = data.shape[0]
    k = data.shape[1] - 1
    df = k * (k - 1) / 2
    chi2_crit = chi2.ppf(1 - alpha, df)
    chi2_val = ((n - 1 - (k - 1) / 2) ** 2) / (k - 1) / ((n - k) * np.mean(data['missing']))
    p_val = 1 - chi2.cdf(chi2_val, df)
    if chi2_val > chi2_crit:
        print(
            'Reject null hypothesis: Data is not MCAR (p-value={:.4f}, chi-square={:.4f})'.format(p_val, chi2_val)
        )
    else:
        print(
            'Do not reject null hypothesis: Data is MCAR (p-value={:.4f}, chi-square={:.4f})'.format(p_val, chi2_val)
        )


In [9]:
little_mcar_test(data, alpha=0.05)

Reject null hypothesis: Data is not MCAR (p-value=0.0000, chi-square=18824.1312)


### `Résultats du test de Little pour l'hypothèse MCAR`

Le test de Little a été effectué pour vérifier si les données sont manquantes complètement au hasard (MCAR). Les résultats montrent que l'hypothèse nulle (les données sont MCAR) est rejetée, ce qui indique que les données ne sont **pas** manquantes complètement au hasard.

- **p-value** = 0.0000 (Une p-value très faible, bien inférieure à 0.05, ce qui signifie que l'hypothèse nulle est rejetée)
- **Chi-square** = 18824.1312 (Valeur du test chi carré)

Ces résultats suggèrent que les données manquantes ne sont probablement pas dues à un hasard pur, mais qu'elles pourraient être influencées par un autre facteur ou mécanisme spécifique. Cela signifie qu'il est nécessaire d'explorer davantage les causes des données manquantes avant de choisir une méthode d'imputation appropriée.

### `Méthode d'imputation bayésienne`

Étant donné que les données ne semblent pas être manquantes complètement au hasard, une approche adaptée pour imputer les valeurs manquantes pourrait être l'imputation bayésienne. Cette méthode permet de prendre en compte l'incertitude associée aux données manquantes en utilisant un modèle probabiliste.

